Component presenter class

tombeynon opened the pull request:
First off - loving MountainView. The front-end team at the agency I work for went nuts over it, but we did find a lot of presentation logic was ending up in the erb files. We usually end up with a large presentation layer in our projects so I figured we could extend MountainView to handle a lot of this logic, and keep the f'enders happy too. 

I noticed this had been raised recently (https://github.com/jgnatch/mountain_view/issues/23) so decided to see if I could come up with a solution you guys would want to merge into master. I've kept everything completely backwards compatible, and tried to keep it as simple and contained as possible. See what you think, I'm definitely open to suggestions.

kitop commented:
Hey @tombeynon, thanks for the contribution!! Really good idea. Glad you're enjoying Mountain View :)

Will take a deeper look into this, and update soon!

Meanwhile, would you mind going over Hound's comments in the meantime? They are really simple, but it's nice to maintain a code styling through the project.

kitop commented on test/dummy/config/initializers/mountain_view.rb:
```diff
@@ -1,3 +1,6 @@
 end
+
+# include component classes
+Dir.glob('app/components/**/*.rb') { |f| require_relative "../../#{f}" }
```
I think it'd be possible to add a line to the `MountainView::Engine` to add the components path as a Rails `autoload_path`, so users don't have to add anything to their initializers.

Something in the lines of the following should work: (haven't tested it)
```ruby
# in lib/mountain_view/engine.rb
initializer "mountain_view.append_component_paths" do |app|
  app.autoload_paths += [MountainView.configuration.components_path]
end
```
It'd probably have to include the sub-directories, so Rails doesn't fail to load the files because there's no namespace.

What do you think?

tombeynon replied:
Totally agree, to be honest I'd put that in temporarily and totally forgotten about it. Thanks for the pointer.

kitop commented on lib/mountain_view/presenter.rb:
```diff
@@ -0,0 +1,70 @@
+    self._attributes = {}
+
+    attr_reader :slug, :properties
+    alias_method :props, :properties
```
Is this alias being used anywhere? Does it make sense to keep it?

tombeynon replied:
The component classes end up with a lot of references to the `properties` hash and I found I wanted a shorter alias. I think it does, but maybe I'm just being lazy :wink:

kitop replied:
Ah, I see! I think I'd be fine to remove it, and people can always alias in in their inherited presenters if they want.

kitop commented on lib/mountain_view/presenter.rb:
```diff
@@ -0,0 +1,70 @@
+
+    def partial
+      "#{slug}/#{slug}"
+    end
```
What do you think of naming this method `to_partial_path`? Ref: https://robots.thoughtbot.com/rendering-collections-in-rails 

It's a bit more Rails-y, and that way we can use `render component, locals: component.locals` in the `render_component` method.

tombeynon replied:
Very nice solution, hadn't even thought of that!

tombeynon replied:
Ah, unfortunately Rails does something a bit special with `to_partial_path`. If the controller is namespaced (such as `MountainView::StyleguideController`) then it'll namespace the `to_partial_path` with the controller namespace, resulting in `mountain_view/component/component/`. See the following method..

https://github.com/rails/rails/blob/20559834e4959531f02f85cac433127c2cc52b20/actionview/lib/action_view/renderer/partial_renderer.rb#L482

The only way I can see to prevent this behaviour would be to override `partial_path` entirely. Not awful, but maybe overkill for what we're trying to achieve.

kitop replied:
Oh, didn't know about that!
Yeah, overriding `partial_path` is definitively overkill. As is, `partial`, should be perfectly fine then.

Thanks for the research!

kitop commented on test/mountain_view/presenter_test.rb:
```diff
@@ -0,0 +1,48 @@
+    assert_equal add_properties(properties), presenter.locals
+  end
+
+  def test_attributes
```
`attributes` is a private method, which causes this test (and another one below) to raise an error. In general, private methods should not be tested. 
Is there a reason this is tested? Was `attributes` public before?

tombeynon replied:
Ah yes, I hastily made the `attributes` method private in a recent commit, it interferes if you pass an attributes key in `render_component`. I don't think this needs to be a public method so I'll remove the broken tests

kitop commented on lib/mountain_view/presenter.rb:
```diff
@@ -0,0 +1,70 @@
+
+    def attributes
+      @attributes ||= properties.keys.inject({}) do |sum, name|
+        sum[name] = {}
```
Why is the default a hash here?
Does it break stuff if we just don't provide a value or make it `nil`?

kitop commented:
@tombeynon Thanks again! Really nice PR, I like the idea and the implementation!!!

Left a few comments through the code, would really like to know what you think about them.

tombeynon commented:
@kitop really glad you like it! I'll reply to your comments and get the Hound violations fixed ASAP.

tombeynon commented:
So now Hound is being quiet (hopefully) and everything is a bit tidier...

We've come up against a couple of obstacles when we've been using MountainView on a large project that I'm hoping we can find a solution for. I think these could tie in with this PR so I'll keep the discussion here for now, but happy to move to an issue or email if you'd rather. 

The main problem is actually with the styleguide. We're finding using a YAML structure to populate the component data quite restrictive - in a lot of cases we want to pass an object but can't stub it using the YAML files. 

There's two main examples of this causing an issue;

**Passing a model instance**
This could be fixed relatively easily by creating an OpenStruct out of the stubbed data and passing that to the component. That way the properties could be accessed using both dot notation and `#[]`. We could tie it in with the Presenter class by introducing an `:as` config when setting up the attributes so we don't have to cast _all_ stubs to OpenStructs, but it would be good to get your opinions on this

```
attribute :banner, as: :object    # stub data is converted to OpenStruct
```  

**Passing a form builder instance**
This could be tough to solve. Often a component is used to represent form fields, without holding the entire form. In this case, we need to pass a form builder instance (such as the one returned from `form_for`), but can't stub it for the styleguide. I haven't really got a solution for this yet..

kitop commented:
@tombeynon thanks a lot!!
Will review again, and merge soon!

Regarding your issue, I totally understand. I was thinking of something similar recently.

There are 2 things I can think of:
- PR #27 will help, as you will be able to add a meta description to the component styleguide. If you want to give it a look and give some feedback, it would be more than welcome!
- What I've been doing is creating kind of param objects and adding them a method called `to_hash`, using that to call `render_component "foo", obj.to_hash` I know that doesn't exactly solve the problem, but it may be helpful.

Another idea would be to introduce marshalized objects in the stubs hash, I haven't tried that and don't know how it'd work, but if it's simple objects, it may be possible. As for form objects, it gets a bit more complex.
How would you like to use the components in that case?

That's definitively for another feature/PR and, as usual, all ideas are welcome!

jgnatch commented:
Hey @tombeynon thanks for your contribution :grinning: 


Completely unrelated: Is your mountain view generated styleguide is public by any chance? I would love to check it out to see how people are using it and maybe link some extra demos.

Thanks again!

tombeynon commented:
@kitop I'll see how the rest of this project goes to decide how best to approach this issue. Will get back to you in a new issue. 

@jgnatch Happy to help! Unfortunately it's not public currently since it's still in heavy development. I'll see if I can give you access when we're done, but since it's a client site I don't think you'd be able to use it for demos I'm afraid. I'll use it on other projects sooner or later though and will let you guys know if I have anything public you can use.

tombeynon commented:
I'm definitely finding some limitations with this implementation. The biggest is due to the way I've exposed the component attributes to the view. These attributes are converted into a hash and passed to the view as locals, but realistically the component wants to have access to component methods, not just variables. This is particularly noticeable when you want to expose an attribute name with a `?` at the end (e.g. `hidden?`) - this will raise an error since a hash key can't contain a `?`. Might be better to utilise PartialRenderer directly, but I need to research this further

kitop commented:
@tombeynon I see what you're saying, and completely agree. We could give access to the object itself, I think it could be worth a try. But then again, we run into the issue of how to deal with that when displaying the styleguide via stubs.

For the `hidden?` thing, doing `:"hidden?"` should work in that case—although it looks bad.

Didn't get the part of
> Might be better to utilise PartialRenderer directly, but I need to research this further
What do you mean there?

tombeynon commented:
@kitop Yea good catch, but I think it's a symptom of a bigger problem really. You expect to expose a method, not a key/value pair in a hash.

Take a look at https://github.com/tombeynon/mountain_view/commit/1cf37a49343da7d267e1dcba378b4b39185eb027 for an alternative way of exposing these methods. I've essentially injected some methods into the view context for the component, which then allows you to access them in the view. It's actually quite elegant.. I need to spend a little more time on it but I think that's the way to go. It does work if you want to test it.

kitop commented:
@tombeynon Interesting! Pretty elegant :)
I like that approach. Just curious, why is presenter inheriting from `SimpleDelegator`?
The runtime object extending by `context.extend ViewContext` + runtime delegation could be a bit costly performance-wise. But really liking where this is headed.

Do you think it's better to wait for that instead of merging this one?

There's one small issue with this current implementation, it fails for Rails 3.2: https://travis-ci.org/jgnatch/mountain_view/jobs/88385844
I'd be great to continue support for older Rails versions, but if that's a big impediment, we can adopt a stricter version policy.

Thanks again for all your help!!!

tombeynon commented:
@kitop SimpleDelegator was just a lazy implementation to get it working. I think it'll probably work better by defining the property methods on the presenter using `define_method`, but I need to have a think about that. Good to know about the performance hit, I hadn't really considered that. 

I definitely think it's worth waiting for me/us to tie this down first. I don't think the current implementation is quite right, and I'm still tweaking it as I'm using it. I would like to get this finished ASAP, but are you guys working to a timeline with this at all? I hopefully should be done with it pretty soon. I'll make sure all tests pass too.

kitop commented:
@tombeynon no hurry! No timeline for this. I believe that taking the time to figure out properly how we are using it is valuable, so we don't end up changing the public API a lot and worrying about keeping backwards compatibility. We can iterate on a branch, and once we're happy, merge it.

Thanks a lot again for your contribution!! Keep us posted on any findings/thoughts :)

devnacho commented:
@tombeynon how's it going? Hope you are doing great! 

Did you have time to continue on this PR?

Thanks!!

tombeynon commented:
@devnacho Hi mate, yea very well thanks. Sorry I've been super busy and kinda let this slide. 

I've picked it back up and will hopefully have some updates for you over the next week or so. Hope that's alright.

devnacho commented:
Hi @tombeynon, 

No hurry, there's no problem at all! 

Thanks for picking it back!

tombeynon commented:
Hey @devnacho - I've spent a bit of time on this, and it's changed so much I decided to open a PR to this PR.. https://github.com/tombeynon/mountain_view/pull/4

Let me know what you think!

kitop commented:
@tombeynon thanks for working on this!!
That PR looks really good. Would you mind rebasing with master and sending a new PR here with all those changes? Would really help for review and merging.

tombeynon commented:
@kitop @devnacho Rebased and fixed any Hound issues. I've also updated the Readme a little more. Happy to create a new PR (or squash some commits) if you want this one cleaned up a bit more. 

Let me know your thoughts on how this all works. I've still got a couple more suggestions (I think it would work a lot better to rename Presenter to Component, so want your opinions on that), but overall I'm pretty happy with where this is.

devnacho commented:
@tombeynon thank you very much for the PR! :grinning: 

Code looks good.

@kitop something to comment or we merge it?

kitop commented:
@tombeynon this looks fantastic!! Thank you so much!!

Yes, rebasing would be nice, if you don't mind.

Happy to hear the rest of the suggestions. I agree that renaming `Presenter` to `Component` can help, but we need to rename the existing `Component` class. Maybe that can be done in a separate PR before bumping the version.

tombeynon commented:
@kitop I definitely think Component would be better, I find myself using it accidentally since it makes more sense semantically. Personally I think the current Component class could be in a `Styleguide` namespace, or simply named StyleguideComponent, but agree it can be a separate PR. Worth getting this in!

Apart from that, I guess my only other concern is whether the README is clear enough about these changes. If you guys are happy, so am I.

I need to squash some of the commits (rebased already), but am currently without a computer, so I'll sort that on Monday at the latest.

kitop commented:
@tombeynon no hurries. And thanks again for all the work.

I think the README is clear, but I'm a bit biased as I've been following the PR and reading the code. I'll have a few colleagues read through it, and if any suggestions I'll let you know or we can update.
